test/jest_integration/saml_auth.test.ts

```diff
@@ -288,6 +288,8 @@ describe('start OpenSearch Dashboards server', () => {
 
     await driver.get('http://localhost:5601/app/opensearch_dashboards_overview#/');
 
+    await driver.wait(until.elementsLocated(By.xpath(signInBtnXPath)), 10000);
```

Comment:
Should we always be waiting for the element before attempting to interact with it? seems like we should wrap the calls to the drive to work in this way - also seems like a bug in how we are using selenium if this is the case.

Reply:
It also has some button clicking behaviors for line: 242, 254 and 272.. should we also locate them before we click that?

Reply:
Did some reading up on selenium and its practices.

[Hard] How we 'wait' for an element to be available should likely be different than the condition to find it in the DOM. .e.g. wait for the sign in button to be _visible_ and _clickable_ then find the sign in button via the DOM and click on it.

Implementing waiters to be reliable might require identifying other DOM elements around the one to be interacted with based on the page loading patterns.  E.g. maybe before we interact with the sign in experience we should make sure a loading spinner isn't drawn, or the pane that holds the sign in button is visible, e


[Easy!] We could ditch all of these wait conditions and use an implicit-wait [1], this will force waiting an amount of time between any actions to ensure the page has stabilized and might result in easier to author test code.


I'm kind of baised to try the easy way first, what do you think?

[1] https://www.selenium.dev/documentation/webdriver/waits/#implicit-wait

Reply:
the implicit wait seems to be a trade-off on finding the correct "wait (number)" time to ensure tests are speedy enough but at the same time, we don't spend too much time waiting. Having said that if this is the only place that we are to introduce implicit wait, I do not mind. But if it affects multiple files, then waiting for page elements to be loaded would be the correct solution.